packages/shell-api/src/shell-internal-state.ts

```diff
@@ -0,0 +1,238 @@
+        {
+          name: '.exit',
+          description: 'shell-api.help.help.exit'
+        }
+      ]
+    });
+  }
+
+  /**
+   * Prepare a `contextObject` as global context and set it as context
+   * Add each attribute to the AsyncWriter also.
+   *
+   * The `contextObject` is prepared so that it can be used as global object
+   * for the repl evaluationi.
+   *
+   * @note The `contextObject` is mutated, it will retain all of its existing
+   * properties but also have the global shell api objects and functions.
+   *
+   * @param {Object} contextObject - contextObject an object used as global context.
+   */
+  setCtx(contextObject: any): void {
+    // Add API methods for VSCode and scripts
+    contextObject.use = this.use.bind(this);
+    contextObject.show = this.show.bind(this);
+    contextObject.it = this.it.bind(this);
+    contextObject.help = this.help.bind(this);
+    contextObject.toIterator = toIterator;
+    contextObject.print = async(arg) => {
+      if (arg.toReplString) {
+        console.log(await arg.toReplString());
+      } else {
+        console.log(arg);
+      }
+    };
+    contextObject.printjson = contextObject.print;
+    Object.assign(contextObject, ShellBson);
+    contextObject.rs = new ReplicaSet(this.currentDb.mongo);
+    contextObject.sh = new Shard(this.currentDb.mongo);
+
+    // Add global shell objects
+    this.asyncWriter.symbols.initializeApiObjects({
+      db: signatures.Database,
+      rs: signatures.ReplicaSet,
+      sh: signatures.Shard,
+      Mongo: signatures.Mongo
+    });
+
+    try {
+      Object.defineProperty(contextObject, 'db', {
+        set: (newDb: any) => {
+          if (newDb.shellApiType && newDb.shellApiType() !== 'Database') {
+            throw new MongoshInvalidInputError('Cannot reassign \'db\' to non-Database type');
+          }
+          this.currentDb = newDb;
+          contextObject.rs = new ReplicaSet(this.currentDb.mongo);
+          contextObject.sh = new Shard(this.currentDb.mongo);
+          this.fetchConnectionInfo();
+          return newDb;
+        },
+        get: () => {
+          return this.currentDb;
+        }
+      });
+    } catch (e) { // java shell, can't use getters/setters
+      console.log(`in java shell, platform: ${this.initialServiceProvider.platform}`);
+      contextObject.db = this.currentDb;
+    }
+
+    contextObject.Mongo = async(uri?, options?): Promise<Mongo> => {
+      if (
+        this.initialServiceProvider.platform !== ReplPlatform.CLI
+      ) {
+        throw new MongoshUnimplementedError(
+          `new Mongo connection are not supported for current platform: ${ReplPlatform[this.initialServiceProvider.platform]}`
+        );
+      }
+      const mongo = new Mongo(this, uri, options);
+      await mongo.connect();
+      return mongo;
+    };
+
+    this.messageBus.emit(
+      'mongosh:setCtx',
+      { method: 'setCtx', arguments: { db: contextObject.db } }
+    );
+  }
+
+  public emitApiCall(event: {
```

Comment:
This doesn't feel like belonging here somehow, I'd extract this in a separate utility class, can be an ad-hoc decorator for `messageBus` which holds all of the `emit` methods:

```
ShellApiEventsEmitter {
   constructor(bus) {}

   private emitApiCall()
   public  emitCollectionApiCall()
   public  emitDatabaseApiCall()
   public  emitCommandApiCall()
}
```
, just the `emitApiCall` function or

Reply:
I moved the other methods out, but this one I kept because it uses the bus. The alternative would be to have a method in the `helpers.ts` file that accepted a bus as the first argument, but that doesn't really change anything except make it slightly more annoying to call from the shell api classes

Reply:
Sounds good.